docs/proposals/005_carbon_loading_mechanism.md

```diff
@@ -0,0 +1,252 @@
+           +--------------------------+          +--------------------------+
+                          |                                     |
+                          v                                     v
+           +--------------------------+          +--------------------------+
+           | Deployment: extension-xyz|          | Deployment: extension-abc|
+           | Image: xyz:latest        |          | Image: abc:latest        |
+           +--------------------------+          +--------------------------+
+```
+
+### Key Concepts
+
+- Extensions are hosted externally and loaded dynamically at runtime.
+- A manifest file lists the extensions and their versions available for loading.
+- Internal and external extensions are bundled and deployed to a designated CDN.
+- External extensions are bundled and deployed to a predefined CDN.
+- The shell application references the CDN URL in the manifest file for seamless integration and dynamic loading at runtime.
+- Kubernetes setup for deploying a shell app and multiple extensions as Docker images, with seamless internal communication.
+
+### Pros and Cons
+
+- **Pros**:
+
+  - **Lightweight Shell App**: Keeps the shell application lightweight by utilizing resources hosted externally.
+  - **Centralized Management**: A single manifest file simplifies the management of extensions by providing centralized information for runtime loading.
+  - **Fault Isolation**: Faulty extensions do not affect the deployment of others, allowing independent rollbacks.
+  - **Dogfooding**: Utilizing your own Docker images for extensions deployed in Kubernetes not only demonstrates confidence in your ecosystem but also validates and tests the artifacts hosted in GHCR, ensuring reliability and alignment with your offered solutions.
+
+- **Cons**:
+  - **Latency Concerns**: Using CDNs instead of Kubernetes-hosted deployments may introduce latency due to external dependencies, potentially affecting user experience.
+  - **Hosting Complexity**: Developers of external extensions must manage their own hosting and deployment to a CDN, adding complexity to the development lifecycle.
+  - **Network Reliability**: The shell application must handle potential network failures when retrieving extensions from external sources.
+  - **CORS Challenges**: Cross-origin resource sharing (CORS) configurations may require careful setup, particularly for external extensions, potentially leading to debugging and integration complexities.
+  - **Internal Extensions Hosting**: Hosting internal extensions would require setting up dedicated asset servers and deployment pipelines. Errors in an extension might necessitate redeploying the host app to update the manifest or enforcing always loading the latest version.
+  - **Kubernetes Setup**: Deploying extensions in Kubernetes demands a more complex setup and ongoing maintenance, which could increase operational overhead.
```

Comment:
- Could be an overkill/over-engineered approach if we are not sure when or whether we will actually have external  extensions to support at all or not.

Reply:
It’s actually not that complicated, but it does require some Kubernetes knowledge and a POC. Once we have an image, like the one we create for Supernova, we need to reference it in a separate deployment, similar to this:  
https://github.com/cloudoperators/greenhouse/blob/main/charts/dashboard/templates/deployment.yaml  
Additionally, we’ll need to create a service:  
https://github.com/cloudoperators/greenhouse/blob/main/charts/dashboard/templates/service.yaml
Once the service is created, we can reference it in the manifest.